### ISSUE TITLE
Q: How i can serve other non-http services via nginx-proxy, like a SMTP or FTP services under a specific virtual hostname?

### ISSUE THREAD

**Issue by chayder** [nginx-proxy/nginx-proxy#1598]
Hello, 

First, i sorry for write here, Google's blog of project are unavailable now...

I'm trying to deploy any non HTTP services that runs over differents ports, like a STMP service running in 25 port, but default.conf nginx file is only generated for http and https trafic.

How i can configure docker nginx-proxy for available this services by reverse proxy?

NOTE: my non http service is running with admin web views in usual http ports also, i need extend the configuration for this virtual hostname for aditional ports and protocols.

I gratefull all help that can bring me.

**Comment by tkw1536** [comment]
TL;DR: nginx-proxy can only delegate for http(s) backends only. 

As far as I am aware, the concept of hostname is a http(s) specific concept.

Protocols built directly on top of tcp or udp do not know about hostnames. All the server sees is the client and server ip address(es) and port(s).  Hostnames are resolved using DNS, and are not made available to the server. As such it is usually not possible for the server to know which "hostname" is being contacted. 

There are only two ways I can think of around this:
- Use different public IP Addresses / Port combinations per service domain. This can be achieved using standard docker networking but requires more than one public ip and/or clients to use non-standard ports. 
- Pick a protocol where the hostname is specified in the first handshake. Then the server can intercept thid hostname and internally forward. This will require protocol specific code. 

For your specific case, the best you could probably do is is the first option, that is for each hostname pick a dedicated port and ip to use for smtp. And then expose the ports for each container using standard docker arguments (`-p $IP:25:25`) etc.

**Comment by pedro2555** [comment]
Check haproxy or nginx plus, both support TCP proxy and load balacing. Nignx plus is not free to use.

_Hostnames in http are a header, the `Host` header._
